docker/Dockerfile

```diff
@@ -73,8 +84,8 @@ RUN apt-get update && \
 	 	echo "INFO: Building Release: ${RELEASE}"; \
 	fi && \
 	RELEASE_CONCAT=$(echo "${RELEASE}" | tr / -); \
-	wget --no-check-certificate -O source.zip https://github.com/"${GITHUB_ACCOUNT}"/"${GITHUB_REPOSITORY}"/archive/"${RELEASE}".zip && \
-	unzip source.zip && \
+	curl -s -L https://github.com/"${GITHUB_ACCOUNT}"/"${GITHUB_REPOSITORY}"/archive/"${RELEASE}".zip > source.zip && \
```

Comment:
@jason-fox is there any reason to download the sources of the app and not use the context sources from where this image is being builded?

I mean, use the `COPY` native function of the Dockerfile and change the build context on the build hook to the root of the repo.

Reply:
Annoyingly `COPY` can't pull from the parent directory. Therefore if the `Dockerfile` isn't in root you have a problem. If the `Dockerfile` is in root then the `README` in the same directory is used on DockerHub - so you still have a problem with broken links in the documentation.

If you think about it, there really is no reason why a `Dockerfile` needs to be in the same repo - many multistage builds start with a _pull external sources from X_ pattern. This happens to use GitHub.

Of course it would be nice to be able to create and build locally with a `COPY` but I've never found a way to do it consistently without amending the base `Dockerfile`.

Reply:
I agree with you, there are a lot of cases where pulling external sources is necessary, but I thing this is not the case, mainly because, allways the image is builded with the sources that actually triggered the dockerhub hook. And of course, it's simpler to use, build and understand.

To build the image you can change the build context on the build hook, take [fiware-cygnus Dockerfile](https://github.com/telefonicaid/fiware-cygnus/blob/master/docker/cygnus-ngsi/hooks/build#L19) as an example. As you can see there, on the build hook we change to the root path by doing 

`cd ../../`

In that case we know for sure that the root of the repo is two folders before the Dockerfile path, so we change the current path before building the image. 

It's not fancy but it works fine, and we know for sure how many folders away is the root path of the repo, so we can change the hook in case the Dockerfile location changes.

Reply:
Unfortunately this is not always true. For the TEF Build on DockerHub you are correct, but as a third-party developer looking to amend or contribute to Perseo, I have no access to your build set-up. Indeed I'm probably looking at fiddling with the codebase and maybe raising a PR. 

The point is that my local code base can easily become inconsistent if I don't create from a recognised source of truth (at least TEF mainline by default). If I fork I can pull from my own branched changes and I can also recreate stable tagged releases by setting the build arguments.

```console
docker build -t perseo . --build-arg DOWNLOAD=latest
```

-   You can alter this to obtain the last **stable** release run this `Dockerfile` with the build argument
    `DOWNLOAD=stable`

```console
docker build -t perseo . --build-arg DOWNLOAD=stable
```

-   You can also download a specific release by running this `Dockerfile` with the build argument `DOWNLOAD=<version>`

```console
docker build -t perseo . --build-arg DOWNLOAD=1.7.0
```

## Building from your own fork

To download code from your own fork of the GitHub repository add the `GITHUB_ACCOUNT`, `GITHUB_REPOSITORY` and
`SOURCE_BRANCH` arguments (default `master`) to the `docker build` command.

```console
docker build -t perseo . \
    --build-arg GITHUB_ACCOUNT=<your account> \
    --build-arg GITHUB_REPOSITORY=<your repo> \
    --build-arg SOURCE_BRANCH=<your branch>
    --target=distroless
```


Basically this aligns perseo with the methods used by the IoT Agents

Reply:
I see. Regardless of the TEF setup, I can't see the advantage of building the image using args rather than using a different repository/branch/account.

From your point of view. What do you think would be the advantage of using those args than checking out to a different branch or fork?

I mean, using the example of the version 1.7.0. You could run:

```bash
git checkout release/1.7.0
cd .. # To change to the root of the perseo repository. Assuming you are on the docker path of this repo.
docker image build -t perseo -f docker/Dockerfile . # In the case you are on a fork, then you should know where the Dockerfile is located
```

This way seems more native to me.

Reply:
But what if I want to update the baseline image from Ubuntu 18.04 to 20.04? What if I want to create a secure distroless build of 1.7.0, but I don't want to update the source code, just the baseline image?

If I'm deploying new updates I want to minimize the number of moving parts so I can test that change in isolation. I may want  to treat the containerization of the `Dockerfile` and the code version independently.

The idea behind the `build-args` is to give developers flexibility to do what they want (which may differ from your personal workflow).  Of course `COPY .` remains as an option.

Reply:
That's part of my point actually. By doing downloading the source code on the dockerfile the `builder` stage must be a debian based image. For example, using alpine as base image would fail. 

```bash
RUN \
	apt-get update && \
	# Ensure that unzip is installed prior to downloading
	apt-get install -y --no-install-recommends unzip ca-certificates curl && \
```

That's the advantage of using the native `COPY` function. It would work on any base distro. And then again, you can allways checkout your workflow to the branch you want to make changes and build from that context.

Anyways. @fgalan & @AlvaroVega what do you think?

Reply:
> `builder` stage must be a debian based image.

Yes, the `builder` stage is however thrown away and not part of your final image. Same would be true of any MUSL based container needing MUSL libs. If you want multistage from alpine you split getting the source code from creating an app - that would be a separate PR of course.

Reply:
@IvanHdzC - this does open a valid question as to which other container variants could/should be supported by the `Dockerfile` - Alpine? RedHat UBI? Additional Architectures beyond ARM? etc. etc.

Reply:
This has gone from the original point. Mi original question was to know why the source code is being downloaded on the dockerfile on the builder stage. I'm not questioning the upgrade to multi stage builds on the dockerfile (Which I think it's a great idea). 

I thing we must use native docker/dockerfile instructions in order to prevent failure. From my point of view the dockerfile should allways build from current context. Just as  fiware-cygnus does. If you want to build from a previous version, you can allways change your workflow and build from that context. 

About the aditional architectures I thing with the current builds it's just fine. As I mentioned, my original question was about the build stage. However @fgalan @AlvaroVega what do you thing about this.

Reply:
After talking this topic with @fgalan and @AlvaroVega, we've decided to accept this PR as it is. 

However we have oppened an issue to take a deeper look into this topic. You can find it and participate on the issue #498